File: Desktop/LED Display Control Example/LedDisplayExample/Model/LedDisplay.cs
/**
 ******************************************************************************
 * @file    LED Display Control Example/Model/LedDisplay.cs
 * @author  Adrian Wojcik  adrian.wojcik@put.poznan.pl
 * @version V1.0
 * @date    07-May-2021
 * @brief   LED display controller: LED display model
 ******************************************************************************
 */

using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Windows.Media;

namespace LedDisplayExample.Model
{
    public class LedDisplay
    {
        public readonly int SizeX = 8;
        public readonly int SizeY = 8;

        public byte ActiveColorA    //!< Active color Alpha components
        {
            get => (byte)((_activeColorR + _activeColorG + _activeColorB) / 3);   
        }

        private byte _activeColorR; 
        public byte ActiveColorR    //!< Active color Red components
        {
            set => _activeColorR = value;
        }

        private byte _activeColorG;
        public byte ActiveColorG    //!< Active color Green components
        {
            set => _activeColorG = value;
        }

        private byte _activeColorB;
        public byte ActiveColorB    //!< Active color Blue components
        {
            set => _activeColorB = value;
        }

        public Color ActiveColor    //!< Active color in ARG format
        {
            get => Color.FromArgb(ActiveColorA, _activeColorR, _activeColorG, _activeColorB);
        }

        public readonly Color OffColor;   //!< 'LED-is-off' color in Int ARGB format

        private UInt16?[,,] model;

        /**
         * @brief Default constructor
         */
        public LedDisplay(int offColor)
        {
            model = new UInt16?[SizeX, SizeY, 3];
            OffColor = Color.FromArgb((byte)(offColor >> 24), (byte)(offColor >> 16), (byte)(offColor >> 8), (byte)offColor);
            _activeColorR = OffColor.R;
            _activeColorG = OffColor.G;
            _activeColorB = OffColor.B;
            ClearModel();
        }

        /**
         * Conversion method: LED x-y position to position/color data in JSON format
         * @param x LED horizontal position in display
         * @param y LED vertical position in display
         * @return Position/color data in JSON format: [x,y,r,g,b] (x,y: 0-7; r,g,b: 0-255)
         */
        private JArray IndexToJsonArray(int x, int y)
        {
            JArray array = new JArray();
            try
            {
                array.Add(x);
                array.Add(y);
                array.Add(model[x, y, 0]);
                array.Add(model[x, y, 1]);
                array.Add(model[x, y, 2]);
            }
            catch (JsonException e)
            {
                Trace.TraceError(e.Message);
            }
            return array;
        }

        /**
         * @brief Null color check
         * @param x LED horizontal position in display
         * @param y LED vertical position in display
         * @return False if color is Null; True otherwise
         */
        private bool ColorNotNull(int x, int y)
        {
            return !((model[x,y,0] == null) || (model[x,y,1] == null) || (model[x,y,2] == null));
        }

        /**
         * @brief Update display model with active color
         * @param x LED horizontal position in display
         * @param y LED vertical position in display
         */
        public void UpdateModel(int x, int y)
        {
            model[x, y, 0] = _activeColorR;
            model[x, y, 1] = _activeColorG;
            model[x, y, 2] = _activeColorB;
        }

        /**
         * @brief LED display data model clear - fill with all components with Null
         */
        public void ClearModel()
        {
            for (int i = 0; i < SizeX; i++)
            {
                for (int j = 0; j < SizeY; j++)
                {
                    model[i, j, 0] = null;
                    model[i, j, 1] = null;
                    model[i, j, 2] = null;
                }
            }
        }

        /**
         * @brief Generate HTTP POST request parameters for LED display control via IoT server script
         * @return HTTP POST request parameters as Key-Value pairs
         */
        public List<KeyValuePair<string, string>> getControlPostData()
        {
            var postData = new List<KeyValuePair<string, string>>();
            for (int i = 0; i < SizeX; i++)
            {
                for (int j = 0; j < SizeY; j++)
                {
                    if (ColorNotNull(i, j))
                        postData.Add(
                            new KeyValuePair<string, string>(
                                "LED" + i.ToString() + j.ToString(),
                                IndexToJsonArray(i,j).ToString()
                                ));
                }
            }
            return postData;
        }

        /**
         * @brief Generate HTTP POST request parameters for clearing LED display via IoT server script
         * @return HTTP POST request parameters as Key-Value pairs
         */
        List<KeyValuePair<string, string>> clearData;
        public List<KeyValuePair<string, string>> getClearPostData()
        {
            if (clearData == null)
            {
                clearData = new List<KeyValuePair<string, string>>();
                for (int i = 0; i < SizeX; i++)
                {
                    for (int j = 0; j < SizeY; j++)
                    {
                        clearData.Add(
                            new KeyValuePair<string, string>(
                                "LED" + i.ToString() + j.ToString(),
                                "["+i.ToString()+ ","+ j.ToString()+",0,0,0]"
                                ));
                    }
                }
            }
            return clearData;
        }
    }
}

File: Android/LED Display Control Example v2/app/src/main/java/iot/examples/leddisplay/LedDisplay.java
/**
 ******************************************************************************
 * @file    LED Display Control Example/LedDisplay.java
 * @author  Adrian Wojcik
 * @version V1.1
 * @date    04-May-2021
 * @brief   LED display controller: LED display model
 ******************************************************************************
 */

package iot.examples.leddisplay;

import org.json.JSONArray;
import org.json.JSONException;

public class LedDisplay {
    public final int sizeX = 8;
    public final int sizeY = 8;

    private int activeColorR; ///< Active color Red components
    private int activeColorG; ///< Active color Green components
    private int activeColorB; ///< Active color Blue components

    public final int offColor;       ///< 'LED-is-off' color in Int ARGB format

    private final Integer[][][] model = new Integer[sizeY][sizeX][3]; ///< LED display data model

    /**
     * @brief Default constructor
     */
    public LedDisplay(int _offColor) {
        offColor = _offColor;
        activeColorR = (offColor >> 16) & 0xff;
        activeColorG = (offColor >> 8) & 0xff;;
        activeColorB = offColor & 0xff;;
        clearModel();
    }

    /**
     * Conversion method: LED x-y position to position/color data in JSON format
     * @param x LED horizontal position in display
     * @param y LED vertical position in display
     * @return Position/color data in JSON format: [x,y,r,g,b] (x,y: 0-7; r,g,b: 0-255)
     */
    private JSONArray indexToJsonArray(int x, int y) {
        JSONArray array = new JSONArray();
        try {
            array.put(0, x);
            array.put(1, y);
            array.put(2, model[x][y][0]);
            array.put(3, model[x][y][1]);
            array.put(4, model[x][y][2]);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return array;
    }

    /**
     * @brief Null color check
     * @param x LED horizontal position in display
     * @param y LED vertical position in display
     * @return False if color is Null; True otherwise
     */
    private boolean colorNotNull(int x, int y) {
        return !((model[x][y][0]==null)||(model[x][y][1]==null)||(model[x][y][2]==null));
    }

    /**
     * @brief Active color getter
     * @return Active color in Int ARGB format
     */
    public int getActiveColor() {
        return  (getActiveColorA() & 0xff) << 24 | (activeColorR & 0xff) << 16 |
                     (activeColorG & 0xff) <<  8 | (activeColorB & 0xff);
    }

    /**
     * @brief Active color getter: A channel
     * @return Alpha channel value
     */
    private int getActiveColorA() {
        return (activeColorR + activeColorG + activeColorB)/3;
    }

    /**
     * @brief Active color setter: R channel
     * @param c Input color value
     */
    public void setActiveColorR(int c) {
        activeColorR = c;
    }

    /**
     * @brief Active color setter: G channel
     * @param c Input color value
     */
    public void setActiveColorG(int c) {
        activeColorG = c;
    }

    /**
     * @brief Active color setter: B channel
     * @param c Input color value
     */
    public void setActiveColorB(int c) {
        activeColorB = c;
    }

    /**
     * @brief Update display model with active color
     * @param x LED horizontal position in display
     * @param y LED vertical position in display
     */
    public void updateModel(int x, int y) {
        model[x][y][0] = activeColorR;
        model[x][y][1] = activeColorG;
        model[x][y][2] = activeColorB;
    }

    /**
     * @brief LED display data model clear - fill with all components with Null
     */
    public void clearModel() {
        for(int i = 0; i < sizeX; i++) {
            for (int j = 0; j < sizeY; j++) {
                model[i][j][0] = null;
                model[i][j][1] = null;
                model[i][j][2] = null;
            }
        }
    }

    /**
     * @brief Generate HTTP PUT request parameters for LED display control via IoT server script
     * @return HTTP PUT request parameters as JSON array
     */
    public JSONArray getControlJsonArray() {
        int led_n = 0;
        JSONArray jsonArray = new JSONArray();
        for(int i = 0; i < sizeX; i++) {
            for (int j = 0; j < sizeY; j++) {
                if(colorNotNull(i,j)) {
                    try {
                        jsonArray.put(led_n, indexToJsonArray(i, j));
                        led_n++;
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                }
            }
        }
        return jsonArray;
    }
}
